src/auto-evo/Miche.cs

```diff
@@ -0,0 +1,210 @@
+    {
+        Pressure = pressure;
+    }
+
+    public bool IsLeafNode()
+    {
+        return Children.Count == 0;
+    }
+
+    public Miche Root()
+    {
+        if (Parent == null)
+        {
+            return this;
+        }
+
+        return Parent.Root();
+    }
+
+    public IEnumerable<Miche> GetLeafNodes()
+    {
+        return GetLeafNodes(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> GetLeafNodes(List<Miche> nodes)
+    {
+        if (IsLeafNode())
+        {
+            nodes.Add(this);
+            return nodes;
+        }
+
+        foreach (var child in Children)
+        {
+            nodes.AddRange(child.GetLeafNodes());
+        }
+
+        return nodes;
+    }
+
+    public IEnumerable<Species> GetOccupants()
+    {
+        return GetOccupants(new List<Species>());
+    }
+
+    public IEnumerable<Species> GetOccupants(List<Species> occupants)
+    {
+        if (Occupant != null)
+        {
+            occupants.Add(Occupant);
+            return occupants;
+        }
+
+        foreach (var child in Children)
+        {
+            occupants.AddRange(child.GetOccupants());
+        }
+
+        return occupants;
+    }
+
+    public IEnumerable<Miche> BackTraversal()
+    {
+        return BackTraversal(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> BackTraversal(List<Miche> currentTraversal)
+    {
+        currentTraversal.Insert(0, this);
+
+        if (Parent == null)
+            return currentTraversal;
+
+        return Parent.BackTraversal(currentTraversal);
+    }
+
+    public void AddChild(Miche newChild)
+    {
+        Children.Add(newChild);
+        newChild.Parent = this;
+    }
+
+    public void AddChildren(IEnumerable<Miche> newChildren)
+    {
+        foreach (var child in newChildren)
+        {
+            AddChild(child);
+        }
+    }
+
+    public bool InsertSpecies(Species species, SimulationCache cache, bool dry = false)
+    {
+        var scoresDictionary = GetOccupants().Distinct().ToDictionary(x => x, _ => 0.0f);
+
+        scoresDictionary[species] = 0.0f;
+
+        return InsertSpecies(species, scoresDictionary, cache, dry);
+    }
+
+    /// <summary>
+    ///   Inserts a species into any spots on the tree where the species is a better fit than any current occupants
+    /// </summary>
+    /// <returns>
+    ///   Returns a bool based on if the species was inserted into a leaf node
+    /// </returns>
+    public bool InsertSpecies(Species species, Dictionary<Species, float> scoresSoFar, SimulationCache cache, bool dry)
+    {
+        var myScore = Pressure.Score(species, cache);
+
+        // Prune branch if species fails any pressures
+        if (myScore <= 0)
+            return false;
+
+        if (IsLeafNode() && Occupant == null)
+        {
+            if (!dry)
+                Occupant = species;
+
+            return true;
+        }
+
+        var newScores = new Dictionary<Species, float>();
```

Comment:
There's probably some way to write this in a way that reduces memory allocations...

Reply:
Ah, so each so to speak "branch" of the execution path through the tree requires its own scores?

Reply:
Yep

Reply:
If this turns out to be one of the top sources of memory allocations, that's not going to be fun, but for now I'll mark this as resolved.